### PR TITLE
data: Simplify to_data_type

### PR DESCRIPTION
**Pull request by JoeLoser**
### :notebook_with_decorative_cover: Description

Problem:
- `to_data_type` has several nested `std::conditional_t` which is slow
  for compile times and hinders readability.
- `to_data_type` is not DRY with use of `std::is_same` several times in
  one branch.
- `to_data_type` uses `std::false_type` to denote the idea of there
  being no mapping for a given `T` to a vast data type. The intent at
  the point of use in the data constructor isn't obvious due to
  comparing with `std::false_type`.

Solution:
- Make `to_data_type` a function template so we can leverage `if
  constexpr` to simplify the implementation (and speed it up).
- Use `detail::is_any_v` to check if `T` is contained within a list of
  types instead of repeated `std::is_same`.
- Introduce a tag type to denote the invalid type mapping case:
  `invalid_data_type`. This makes it clear from the use in the data
  constructor the intent by comparing with `invalid_data_type` instead
  of `std::false_type`.

### :memo: Checklist

- [X] All user-facing changes have changelog entries.
- [X] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [X] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

File-by-file.

### CODE REVIEW

File: libvast/vast/data.hpp [hunk 1]
```diff
@@ -15,6 +15,7 @@
 #include "vast/data/integer.hpp"
 #include "vast/defaults.hpp"
 #include "vast/detail/operators.hpp"
+#include "vast/detail/type_list.hpp"
```
Comment (mavam) [inline libvast/vast/data.hpp:18]:
```suggestion
```

Is this really needed?

**Review by mavam — COMMENTED** [review]
Interesting. You are using the return type of a generic function to dispatch. I like the approach. Any objections from @tenzir/backend?

**Review by dominiklohmann — APPROVED** [review]
I like this change.

Note that we have plans to convert the remaining types in data that are weak type aliases to strong custom types, like we've done in https://github.com/tenzir/vast/pull/1604 for `integer`.

Once `boolean`, `enumeration`, `count`, `real`, and `string` are converted we can get rid of the `to_data_type` heuristic once and for all.

File: libvast/vast/data.hpp (hunk 1 above)
Comment (JoeLoser) [reply libvast/vast/data.hpp]: No. I meant to include `vast/detail/type_traits.hpp` -- this is just a typo. I just pushed a fix so we include `vast/detail/type_traits.hpp` for `detail::is_any_v` definition.

**Pull request merged**